pkg/trace/sampler/remote_rates.go

```diff
@@ -0,0 +1,218 @@
+package sampler
+
+import (
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/config/remote/service"
+	"github.com/DataDog/datadog-agent/pkg/proto/pbgo"
+	"github.com/DataDog/datadog-agent/pkg/trace/config"
+	"github.com/DataDog/datadog-agent/pkg/trace/pb"
+	"github.com/DataDog/datadog-agent/pkg/trace/watchdog"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// RemoteRates sharded by signature. Allowing an independant feedback
+// loop per signature. RemoteRates adjusts sampling rates communicated
+// to the agent based on the observed traffic. It targets a TPS configured
+// remotely.
+type RemoteRates struct {
+	samplers   map[Signature]*Sampler
+	tpsTargets map[Signature]float64
+	mu         sync.RWMutex // protects concurrent access to samplers and tpsTargets
+
+	exit    chan struct{}
+	stopped chan struct{}
+}
+
+func newRemoteRates(conf *config.AgentConfig) *RemoteRates {
+	if !conf.EnabledRemoteRates {
+		return nil
+	}
+
+	remoteRates := &RemoteRates{
+		samplers: make(map[Signature]*Sampler),
+
+		exit:    make(chan struct{}),
+		stopped: make(chan struct{}),
+	}
+
+	if err := service.NewGRPCSubscriber(pbgo.Product_APM_SAMPLING, remoteRates.subscriberCallback()); err != nil {
```

Comment:
How is this implemented? Does it mean we are keeping connections open to the backend?

Reply:
This is a connection to the core-agent grpc server

Reply:
This is a grpc client to a grpc server in the core agent. It hits a similar server that is used to resolve host names.
Core-agent handles connection to the backend and backend requests.
It pushes to subscribers updates when available.

Reconnection logic to core agent is handled in NewGRPCSubscriber code